### DISCUSSION TITLE
How is phash metric calculated when comparing images?

### DISCUSSION THREAD

**Discussion by alxersov** [ImageMagick/ImageMagick#5191] [Help]
If I run:

```
compare -verbose -metric phash baboon.tiff peppers.tiff NULL:
```

ImageMagick 7.1.0-36 returns:

```
Channel distortion: PHASH
  red: 0.309604
  green: 0.336803
  blue: 0.333855
  all: 0.326981
```

While ImageMagick 6.9.12-51 returns different values:

```
Channel distortion: PHASH
  red: 23.6422
  green: 15.2175
  blue: 48.3749
  all: 87.2346
```

I would like to know, what is the exact formula used in `compare` with perceptual hash metric? Why the results are different in ImageMagick 6 and 7?

How can I calculate those values if I know 42 phash values for both files? 

```
identify -verbose -moments baboon.tiff

Channel perceptual hash:
  Red, Hue:
    PH1: 0.533154, 0.351193
    PH2: 3.52162, 3.052
    PH3: 4.73929, 3.62367
    PH4: 3.77149, 3.79462
    PH5: 8.39265, 8.24478
    PH6: 5.6235, 5.38934
    PH7: 8.07144, 7.51104
  Green, Chroma:
    PH1: 0.478523, 0.263804
    PH2: 3.56258, 2.55898
    PH3: 4.87945, 2.95157
    PH4: 4.95299, 2.82515
    PH5: 9.90383, 5.75631
    PH6: 6.8048, 4.27201
    PH7: 10.285, 6.08724
  Blue, Luma:
    PH1: 0.467656, 0.493031
    PH2: 3.59239, 3.59731
    PH3: 5.09744, 4.87978
    PH4: 4.26879, 4.83728
    PH5: 9.09465, 9.87445
    PH6: 6.9776, 6.89281
    PH7: 9.11046, 9.82144
```

```
identify -verbose -moments peppers.tiff

Channel perceptual hash:
  Red, Hue:
    PH1: 0.54073, 0.192403
    PH2: 4.4014, 1.81032
    PH3: 3.88971, 3.27699
    PH4: 3.79898, 2.5485
    PH5: 7.67239, 5.73009
    PH6: 6.00857, 3.56615
    PH7: 8.09446, 5.53559
  Green, Chroma:
    PH1: 0.436404, 0.36581
    PH2: 2.46433, 3.52255
    PH3: 2.98936, 3.33504
    PH4: 4.52607, 3.12384
    PH5: 8.84538, 6.5714
    PH6: 6.00878, 5.03648
    PH7: 8.30078, 6.45231
  Blue, Luma:
    PH1: 0.199033, 0.450269
    PH2: 2.44932, 2.96879
    PH3: 2.6395, 3.47587
    PH4: 2.8886, 4.34987
    PH5: 6.09574, 8.30335
    PH6: 4.16242, 5.90771
    PH7: 5.68288, 8.64679
```

**Comment by fmw42** [comment]
See http://www.fmwconcepts.com/misc_tests/perceptual_hash_test_results_510/index.html

and the original paper at https://www.naturalspublishing.com/files/published/54515x71g3omq1.pdf

**Reply by alxersov** [reply]
@fmw42 thank you for reply.

The Tests Of Perceptual Hash page mentions:

$$ \sum\_{i=1}^{42}{[h_1(i)-h_2(i)]^2} $$

The original paper suggests:

$$ \sqrt{\sum\_{i=1}^{42}{[h_1(i)-h_2(i)]^2}} $$

The 2016 thread on legacy development forum [Get perceptual hash value for image using command line](https://legacy.imagemagick.org/discourse-server/viewtopic.php?f=2&t=30258) suggests other calculation methods (proportional difference, RMS etc). And it is not clear which method was implemented.

If I calculate sum of the squared differences of all 42 values, I get 77.58133471 which does not match ImageMagick result (87.2346, nor 0.326981). Probably the formula was updated since 2016?

|        | Baboon   | Peppers  | sum of the squared differences |
| :----- | :------- | :------- | ------------------------------ |
| Red    | 0.533154 | 0.54073  | 0.000057395776                 |
|        | 3.52162  | 4.4014   | 0.7740128484                   |
|        | 4.73929  | 3.88971  | 0.7217861764                   |
|        | 3.77149  | 3.79898  | 0.0007557001                   |
|        | 8.39265  | 7.67239  | 0.5187744676                   |
|        | 5.6235   | 6.00857  | 0.1482789049                   |
|        | 8.07144  | 8.09446  | 0.0005299204                   |
| Green  | 0.478523 | 0.436404 | 0.001774010161                 |
|        | 3.56258  | 2.46433  | 1.206153063                    |
|        | 4.87945  | 2.98936  | 3.572440208                    |
|        | 4.95299  | 4.52607  | 0.1822606864                   |
|        | 9.90383  | 8.84538  | 1.120316403                    |
|        | 6.8048   | 6.00878  | 0.6336478404                   |
|        | 10.285   | 8.30078  | 3.937129008                    |
| Blue   | 0.467656 | 0.199033 | 0.07215831613                  |
|        | 3.59239  | 2.44932  | 1.306609025                    |
|        | 5.09744  | 2.6395   | 6.041469044                    |
|        | 4.26879  | 2.8886   | 1.904924436                    |
|        | 9.09465  | 6.09574  | 8.993461188                    |
|        | 6.9776   | 4.16242  | 7.925238432                    |
|        | 9.11046  | 5.68288  | 11.74830466                    |
| Hue    | 0.351193 | 0.192403 | 0.0252142641                   |
|        | 3.052    | 1.81032  | 1.541769222                    |
|        | 3.62367  | 3.27699  | 0.1201870224                   |
|        | 3.79462  | 2.5485   | 1.552815054                    |
|        | 8.24478  | 5.73009  | 6.323665796                    |
|        | 5.38934  | 3.56615  | 3.324021776                    |
|        | 7.51104  | 5.53559  | 3.902402703                    |
| Chroma | 0.263804 | 0.36581  | 0.01040522404                  |
|        | 2.55898  | 3.52255  | 0.9284671449                   |
|        | 2.95157  | 3.33504  | 0.1470492409                   |
|        | 2.82515  | 3.12384  | 0.0892157161                   |
|        | 5.75631  | 6.5714   | 0.6643717081                   |
|        | 4.27201  | 5.03648  | 0.5844143809                   |
|        | 6.08724  | 6.45231  | 0.1332761049                   |
| Luma   | 0.493031 | 0.450269 | 0.001828588644                 |
|        | 3.59731  | 2.96879  | 0.3950373904                   |
|        | 4.87978  | 3.47587  | 1.970963288                    |
|        | 4.83728  | 4.34987  | 0.2375685081                   |
|        | 9.87445  | 8.30335  | 2.46835521                     |
|        | 6.89281  | 5.90771  | 0.97042201                     |
|        | 9.82144  | 8.64679  | 1.379802623                    |
| Sum    |          |          | 77.58133471                    |

**Comment by fmw42** [comment]
I used sum of squared differences in my scripts.

**Reply by fmw42** [reply]
I cannot say exactly how it was implemented in IM. But what I know was stated on my test page

> The hash is created by taking the log of the absolute value of each of the first 7 floating point Hu image moments from each channel using graylevels normalized to the range 0 to 1 (as of IM 6.8.8.5). The log is used to mitigate the large variation in exponents of the moments. The absolute value is used to make the moments insensitive to mirror-like attacks. For color images, there are 42 floating point values. For grayscale images, there are only 7 floating point moments. Therefore, the hash lengths are different for color images and grayscale images. Consequently, color and grayscale images cannot be compared directly. The phash metric computes the sum of the squared differences of all 42 or 7 

You could be seeing a precision issue due to the log in the computation.

**Reply by fmw42** [reply]
There may be an error in Imagemagick for the normalized version which is being looked into right now. Also the definition of the normalized version needs clarification and is being worked on.
